device/transport/amqp/src/amqp_device_method_client.ts

```diff
@@ -71,74 +93,118 @@ export class AmqpDeviceMethodClient extends EventEmitter {
           }
         },
-        'connected': {
+        attached: {
+          _onEnter: (attachCallback) => {
+            attachCallback();
+          },
           sendMethodResponse: (response, callback) => {
             let message = new Message(JSON.stringify(response.payload));
             message.correlationId = response.requestId;
             message.properties.add(methodMessagePropertyKeys.status, response.status);
             this._amqpClient.send(message, this._methodEndpoint, undefined, callback);
           },
+          /*Codes_SRS_NODE_AMQP_DEVICE_METHOD_CLIENT_16_006: [The `onDeviceMethod` method shall save the `callback` argument so that it is called when the corresponding method call is received.]*/
           onDeviceMethod: (methodName, methodCallback) => {
-            this._amqpClient.getReceiver(this._methodEndpoint, (err, methodReceiver) => {
-              if (!this._methodReceiverInitialized) {
-                methodReceiver.on('message', (msg) => {
-                  debug('got method request');
-                  debug(JSON.stringify(msg, null, 2));
-                  const methodName = msg.properties.getValue(methodMessagePropertyKeys.methodName);
-                  const methodRequest = {
-                    methods: { methodName: methodName },
-                    requestId: msg.correlationId,
-                    body: msg.getData()
-                  };
-
-                  debug(JSON.stringify(methodRequest, null, 2));
-                  this.emit('method_' + methodName, methodRequest);
-                });
-                methodReceiver.on('errorReceived', (err) => {
-                  this._fsm.transition('disconnecting');
-                  this._amqpClient.detachReceiverLink(this._methodEndpoint, () => {
-                    this._amqpClient.detachSenderLink(this._methodEndpoint, () => {
-                      this._methodReceiverInitialized = false;
-                      this._fsm.transition('disconnected');
-                      this.emit('errorReceived', err);
-                    });
-                  });
-                });
-                this._methodReceiverInitialized = true;
+            debug('attaching callback for method: ' + methodName);
+            this.on('method_' + methodName, methodCallback);
+          },
+          /*Codes_SRS_NODE_AMQP_DEVICE_METHOD_CLIENT_16_020: [The `attach` method shall immediately call the callback if the links are already attached.]*/
+          attach: (callback) => callback(),
+          detach: (callback) => this._fsm.transition('detaching', callback),
+          forceDetach: () => {
+            /*Codes_SRS_NODE_AMQP_DEVICE_METHOD_CLIENT_16_024: [The `forceDetach` method shall forcefully detach all links.]*/
+            this._senderLink.forceDetach();
+            this._receiverLink.forceDetach();
+            this._fsm.transition('detached');
+          }
+        },
+        detaching: {
+          _onEnter: (forwardedCallback, err) => {
```

Comment:
>forwardedCallback [](start = 21, length = 17)

I'm confused at this point.  The entry handler is getting this forwardedCallback but I don't see how/what the async.each function passes a callback.

Reply:
[https://caolan.github.io/async/docs.html#each](https://caolan.github.io/async/docs.html#each)
each function is executed, for each of those functions the local callback (named callback) is used, then once all functions have called their local callback, the transition is called with the forwarded callback.

---
In reply to: [140033506](https://github.com/Azure/azure-iot-sdk-node/pull/85#discussion_r140033506) [](ancestors = 140033506)